src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleMetadata.cs

```diff
@@ -27,18 +27,24 @@ public static partial class JsonSerializer
             {
                 metadataToWrite = MetadataPropertyName.NoMetadata;
             }
-            else if (state.ReferenceResolver.TryGetOrAddReferenceOnSerialize(currentValue, out string referenceId))
-            {
-                Debug.Assert(referenceId != null);
-                writer.WriteString(s_metadataRef, referenceId);
-                writer.WriteEndObject();
-                metadataToWrite = MetadataPropertyName.Ref;
-            }
             else
             {
-                Debug.Assert(referenceId != null);
-                writer.WriteString(s_metadataId, referenceId);
-                metadataToWrite = MetadataPropertyName.Id;
+
+                string referenceId = state.ReferenceResolver.GetReference(currentValue, out bool alreadyExists);
+
+                if (alreadyExists)
+                {
+                    Debug.Assert(referenceId != null);
+                    writer.WriteString(s_metadataRef, referenceId);
+                    writer.WriteEndObject();
+                    metadataToWrite = MetadataPropertyName.Ref;
```

Comment:
Hmmm, I guess this was named this way because it used to be the return value of an old method that would indicate which metadata property should be written within this context.

https://github.com/dotnet/runtime/pull/2259/files#diff-bdcd9cf5b1e1f2aa5850ad65e488269bR43-R55

Now that we determine the value within the same method and we return it, it does make sense use `writtenMetadataName`.